motto/agents/retrieval_agent.py

```diff
@@ -0,0 +1,116 @@
+        self.collection = self.chroma_client.get_or_create_collection(name=self.collection_name,
+                                                                      metadata={"hnsw:space": "cosine"})
+
+        if not need_load_docs:
+            if self.collection.count() == 0:
+                need_load_docs = True
+                shutil.rmtree(self.db)
+
+        if need_load_docs:
+            self._load_docs()
+
+    def _process_doc(self, file):
+        chunks = {}
+        text = DocProcessor.clear_text(file)
+        chunks['texts'] = DocProcessor.get_text_chunks(text, self.chunk_token_size)
+        chunks['embeddings'] = self.embeddings_getter.get_chunks_embeddings(chunks['texts'])
+        length = len(chunks['texts'])
+        if length > 0:
+            chunks['source'] = [{'source': file}] * length
+        return chunks
+
+    def _load_docs(self):
+        try:
+            with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
+                futures = []
+                if os.path.isfile(self.data_source):
+                    files = [self.data_source]
+                else:
+                    files = [os.path.join(self.data_source, file) for file in  os.listdir(self.data_source)]
+                for file in files:
+                    futures.append(executor.submit(self._process_doc, file=file))
+                i = 0
+                for future in concurrent.futures.as_completed(futures):
+                    chunks = future.result()
+                    length = len(chunks['texts'])
+                    chunks['ids'] = [f"id{i + j}" for j in range(length)]
+                    i += length
+                    self.collection.add(embeddings=chunks['embeddings'], documents=chunks['texts'],
+                                        metadatas=chunks['source'], ids=chunks['ids'])
+        except Exception as ex:
+            if os.path.exists(self.db):
+                shutil.rmtree(self.db)
+            raise RuntimeError(f"RetrievalAgent.__load_docs error: {ex}")
+
+    def __call__(self, messages, docs_count=10):
+        if isinstance(messages, str):
+            text = messages
+        else:
+            try:
+                text = list(map(lambda m: m['content'], messages))
+                text = '\n'.join(text)
+            except Exception as ex:
+                raise TypeError(f"Incorrect argument for retrieval-agent: {ex}")
+        embeddings_values = self.embeddings_getter.get_embeddings(text)
+        context = self.collection.query(query_embeddings=embeddings_values, n_results=docs_count)
+        docs = context["documents"][0]
+        res = ""
+        prev = ""
+        for doc in docs:
+            next = doc.replace('"',"'")
+            if next not in res:
+                res += next + "\n"
+                # metta truncates the results, the maximum length of result is 2276, so I do not add chunks if the final string becomes longer than 2276
```

Comment:
This seems to be a relatively major problem.  self.max_length = 2270 seems to be rather random constant. It leads to unexpected behavior. cc : @Necr0x0Der